lib/db/set.go

```diff
@@ -80,25 +80,33 @@ func NewFileSet(folder string, fs fs.Filesystem, db *Lowlevel) *FileSet {
 	}
 
 	if err := s.meta.fromDB(db, []byte(folder)); err != nil {
-		l.Infof("No stored folder metadata for %q: recalculating", folder)
-		if err := s.recalcCounts(); backend.IsClosed(err) {
-			return nil
-		} else if err != nil {
-			panic(err)
-		}
-	} else if age := time.Since(s.meta.Created()); age > databaseRecheckInterval {
+		l.Infof("No stored folder metadata for %q; recalculating", folder)
+		goto recalcMeta
+	}
+
+	if metaOK := s.verifyLocalSequence(); !metaOK {
+		l.Infof("Stored folder metadata for %q is out of date after crash; recalculating", folder)
+		goto recalcMeta
+	}
+
+	if age := time.Since(s.meta.Created()); age > databaseRecheckInterval {
 		l.Infof("Stored folder metadata for %q is %v old; recalculating", folder, age)
-		if err := s.recalcCounts(); backend.IsClosed(err) {
-			return nil
-		} else if err != nil {
-			panic(err)
-		}
+		goto recalcMeta
+	}
+
+	return &s
```

Comment:
This just made me blink, then think about the goto, think again, and then get it. Personally I'd rather use a function variable defined above like

```
recalc := func() *FileSet {
	if err := s.recalcMeta(); backend.IsClosed(err) {
		return nil
	} else if err != nil {
		panic(err)
	}
	return &s
}
```

and replace the gotos with calling and returning that.